src/components/tx/SignOrExecuteForm/index.tsx

```diff
@@ -129,11 +133,18 @@ const SignOrExecuteForm = ({
       return
     }
 
+    // For Safe Apps:
+    // - Avoid redirection
+    // - Dispatch an event for return the safeTxHash to the caller
+    if (requestId) {
+      txDispatch(TxEvent.SAFE_APPS_REQUEST, { txId: id, requestId })
+      return
```

Comment:
I'd suggest not to return here and instead let if call `onSubmit` further down.

To avoid redirecting, just add a requestId check to the if below.

Reply:
I prefer to take control of the modal because when user press the cancel button or the cross icon i need to send a message to the Safe App

Reply:
You can still do it onSubmit, can't you? I don't like that the Safe App request behaves different than other types of modals. IMHO they should all call onSubmit on the final step.

Reply:
Ok, refactored as suggested leaving the SignOrExecuteForm as before but with the id in the Submit. Adjusted other components using it

Reply:
Cool, thanks Yago!